src/api/channel.rs

```diff
@@ -0,0 +1,649 @@
+    Self { sender, limit, count: 0 }
+  }
+  pub fn try_send(
+    &mut self, data: RcData,
+  ) -> Result<(), TrySendError<RcData>> {
+    if self.limit < self.count {
+      Err(TrySendError::Full(data))
+    } else {
+      let r = self.sender.try_send(data);
+      if r.is_ok() {
+        self.count += 1;
+      }
+      r
+    }
+  }
+  pub fn send(&mut self, data: RcData) -> Result<(), SendError<RcData>> {
+    if self.limit < self.count {
+      Err(SendError(data))
+    } else {
+      let r = self.sender.send(data);
+      if r.is_ok() {
+        self.count += 1;
+      }
+      r
+    }
+  }
+  pub fn len(&self) -> usize {
+    self.sender.len()
+  }
+  pub fn is_empty(&self) -> bool {
+    self.sender.is_empty()
+  }
+
+  // TODO: proxy more methods
+}
+
+/// Endpoint to receive current-pass statistics data
+pub struct RcDataReceiver(Receiver<RcData>);
+
+impl RcDataReceiver {
+  pub fn try_recv(&self) -> Result<RcData, TryRecvError> {
+    self.0.try_recv()
+  }
+  pub fn recv(&self) -> Result<RcData, RecvError> {
+    self.0.recv()
+  }
+  pub fn len(&self) -> usize {
+    self.0.len()
+  }
+  pub fn is_empty(&self) -> bool {
+    self.0.is_empty()
+  }
+  pub fn iter(&self) -> Iter<RcData> {
+    self.0.iter()
+  }
+
+  pub fn summary_size(&self) -> usize {
+    crate::rate::TWOPASS_HEADER_SZ
+  }
+}
+
+pub type PassDataChannel = (RcDataSender, RcDataReceiver);
+
+pub type FrameInput<T> = (Option<Arc<Frame<T>>>, Option<FrameParameters>);
+
+/// Endpoint to send frames
+pub struct FrameSender<T: Pixel> {
+  sender: Sender<FrameInput<T>>,
+  enc: Arc<EncoderConfig>,
+  limit: u64,
+  count: u64,
+}
+
+// Proxy the crossbeam Sender
+//
+// TODO: enforce the limit
+impl<T: Pixel> FrameSender<T> {
+  fn new(
+    limit: u64, sender: Sender<FrameInput<T>>, enc: Arc<EncoderConfig>,
+  ) -> FrameSender<T> {
+    Self { sender, enc, limit, count: 0 }
+  }
+  pub fn try_send<F: IntoFrame<T>>(
+    &mut self, frame: F,
+  ) -> Result<(), TrySendError<FrameInput<T>>> {
+    if self.limit < self.count {
+      Err(TrySendError::Full(frame.into()))
+    } else {
+      let r = self.sender.try_send(frame.into());
+      if r.is_ok() {
+        self.count += 1;
+      }
+      r
+    }
+  }
+
+  pub fn send<F: IntoFrame<T>>(
+    &mut self, frame: F,
+  ) -> Result<(), SendError<FrameInput<T>>> {
+    if self.limit < self.count {
+      Err(SendError(frame.into()))
+    } else {
+      let r = self.sender.send(frame.into());
+      if r.is_ok() {
+        self.count += 1;
+      }
+      r
+    }
+  }
+  pub fn len(&self) -> usize {
+    self.sender.len()
+  }
+  pub fn is_empty(&self) -> bool {
+    self.sender.is_empty()
+  }
+  // TODO: proxy more methods
+}
+
+// Frame factory
+impl<T: Pixel> FrameSender<T> {
+  /// Helper to create a new frame with the current encoder configuration
+  #[inline]
+  pub fn new_frame(&self) -> Frame<T> {
+    Frame::new(self.enc.width, self.enc.height, self.enc.chroma_sampling)
```

Comment:
This will probably need to be changed given the removal of the Frame::new API.

Reply:
It is removed only from the public API, but I had to change rav1e-ch.rs due to that.